File: twosum.cpp
// https://leetcode.com/problems/two-sum
#include <map>
#include <vector>

using namespace std;

class Solution {
public:
	vector<int> twoSum(vector<int>& nums, int target) {
		map<int,int> ns;
		vector<int> v(2);

		for(int i = 0; i < nums.size(); i++) {
			ns[nums[i]] = i+1;
		}
		for(int i = 0; i < nums.size(); i++) {
			if(ns[target-nums[i]] && i != ns[target-nums[i]]-1) {
				v[0] = i;
				v[1] = ns[target-nums[i]]-1;
				return v;
			}
		}
		return v;

	}
};

File: elixir_solutions/lib/twosum.ex
# https://leetcode.com/problems/two-sum

defmodule Twosum do
  def solve(target, list, existent_indexes \\ {0, %{}})

  def solve(_, [], _) do
    {nil, nil}
  end

  def solve(target, [head | tail], {i, existent}) do
    existent_inx = existent[target - head]

    if is_nil(existent_inx) do
      solve(target, tail, {i + 1, Map.put(existent, head, i)})
    else
      {existent_inx, i}
    end
  end
end
